tools/python/boutiques/puller.py

```diff
@@ -25,31 +25,47 @@ def __init__(self, zid, verbose, download, sandbox):
         self.verbose = verbose
         self.download = download
         self.sandbox = sandbox
+        self.no_int = no_int
+        self.cache_dir = os.path.join(os.path.expanduser('~'), ".cache",
+                                      "boutiques")
 
     def pull(self):
         from boutiques.searcher import Searcher
-        searcher = Searcher(self.zid, self.verbose, self.sandbox, None)
+        searcher = Searcher(self.zid, self.verbose, self.sandbox, None, True)
         r = searcher.zenodo_search()
 
         for hit in r.json()["hits"]["hits"]:
             file_path = hit["files"][0]["links"]["self"]
             file_name = file_path.split(os.sep)[-1]
             if hit["id"] == int(self.zid):
                 if self.download:
-                    cache_dir = os.path.join(os.path.expanduser('~'), ".cache",
-                                             "boutiques")
-                    if not os.path.exists(cache_dir):
-                        os.makedirs(cache_dir)
+                    if not os.path.exists(self.cache_dir):
+                        os.makedirs(self.cache_dir)
+                    elif(not self.no_int and
+                         os.path.isfile(os.path.join(self.cache_dir,
+                                                     file_name))):
+                        prompt = ("Found existing file with the same name. "
+                                  "Overwrite? (Y/n) ")
+                        try:
+                            ret = raw_input(prompt)  # Python 2
+                        except NameError:
+                            ret = input(prompt)  # Python 3
+                        if ret.upper() != "Y":
+                            return
                     if(self.verbose):
                         print_info("Downloading descriptor %s"
                                    % file_name)
-                    downloaded = urlretrieve(file_path, os.path.join(cache_dir,
-                                             file_name))
-                    print_info("Downloaded descriptor to " + cache_dir)
+                    downloaded = urlretrieve(file_path,
+                                             os.path.join(self.cache_dir,
+                                                          file_name))
+                    print("Downloaded descriptor to " + self.cache_dir)
                     return downloaded
                 if(self.verbose):
                     print_info("Opening descriptor %s"
                                % file_name)
+                # use the cached file if it exists
+                if os.path.isfile(os.path.join(self.cache_dir, file_name)):
+                    return open(os.path.join(self.cache_dir, file_name), "r")
```

Comment:
If we save the descriptor with the filename matching the doi, do we really need to include the doi in the descriptor itself too? I think it's enough to just have it in the filename. Also I don't really like the idea of modifying a descriptor before publishing it without the user knowing.

Reply:
We modify the descriptor already, it's just that we modify it after publication. It's ambiguous because the user now gets a different descriptor than that published. I think it's useful to add the DOI to the descriptor once published, because the DOI is the only permanent identifier.

Reply:
We only get the DOI *after* publication, so I don't think this is possible. 

From [Zenodo API docs](http://developers.zenodo.org/#depositions): "When you publish your deposition, we register a DOI in DataCite for your upload, unless you manually provided us with one. This field is only present for published depositions."

Reply:
ok, makes sense then.